raft/fuzz_test.go

```diff
@@ -0,0 +1,215 @@
+
+package raft
+
+import (
+	"os"
+	"runtime"
+	"strings"
+	"sync"
+	"testing"
+
+	fuzz "github.com/AdaLogics/go-fuzz-headers"
+	pb "go.etcd.io/etcd/raft/v3/raftpb"
+
+	"go.uber.org/zap"
+	"go.uber.org/zap/zapcore"
+)
+
+var initter sync.Once
+
+func initFunc() {
+	SetLogger(discardLogger)
+}
+
+func getMsgType(i int) pb.MessageType {
+	allTypes := map[int]pb.MessageType{0: pb.MsgHup,
+		1:  pb.MsgBeat,
+		2:  pb.MsgProp,
+		3:  pb.MsgApp,
+		4:  pb.MsgAppResp,
+		5:  pb.MsgVote,
+		6:  pb.MsgVoteResp,
+		7:  pb.MsgSnap,
+		8:  pb.MsgHeartbeat,
+		9:  pb.MsgHeartbeatResp,
+		10: pb.MsgUnreachable,
+		11: pb.MsgSnapStatus,
+		12: pb.MsgCheckQuorum,
+		13: pb.MsgTransferLeader,
+		14: pb.MsgTimeoutNow,
+		15: pb.MsgReadIndex,
+		16: pb.MsgReadIndexResp,
+		17: pb.MsgPreVote,
+		18: pb.MsgPreVoteResp}
+	return allTypes[i%len(allTypes)]
+}
+
+// All cases in shouldReport represent known errors in etcd
+// as these are reported via manually added panics.
+func shouldReport(err string) bool {
+	if strings.Contains(err, "stepped empty MsgProp") {
+		return false
+	}
+	if strings.Contains(err, "Was the raft log corrupted, truncated, or lost?") {
+		return false
+	}
+	if strings.Contains(err, "ConfStates not equivalent after sorting:") {
+		return false
+	}
+	if strings.Contains(err, "term should be set when sending ") {
+		return false
+	}
+	if (strings.Contains(err, "unable to restore config")) && (strings.Contains(err, "removed all voters")) {
+		return false
+	}
+	if strings.Contains(err, "ENCOUNTERED A PANIC OR FATAL") {
+		return false
+	}
+	if strings.Contains(err, "index, ") && strings.Contains(err, ", is out of range [") {
+		return false
+	}
+	// This string is found in raft.go because we change all
+	// occurrences from panic(err) to panic("GOT A FUZZ ERROR").
+	// This is done in build.sh as a simple solution to catch
+	// all the panic(err)'s that would otherwise be fuzz
+	// blockers.
+	if strings.Contains(err, "GOT A FUZZ ERROR") {
+		return false
+	}
+
+	return true
+}
+
+func catchPanics() {
+	if r := recover(); r != nil {
+		var err string
+		switch r.(type) {
+		case string:
+			err = r.(string)
+		case runtime.Error:
+			err = r.(runtime.Error).Error()
+		}
+		if shouldReport(err) {
+			// Getting to this point means that the fuzzer
+			// did not stop because of a manually added panic.
+			panic(err)
+		}
+	}
+}
+
+func silentLogger(options ...zap.Option) *zap.Logger {
+	encoderCfg := zapcore.EncoderConfig{
+		MessageKey:     "msg",
+		LevelKey:       "level",
+		NameKey:        "logger",
+		EncodeLevel:    zapcore.LowercaseLevelEncoder,
+		EncodeTime:     zapcore.ISO8601TimeEncoder,
+		EncodeDuration: zapcore.StringDurationEncoder,
+	}
+	core := zapcore.NewCore(zapcore.NewJSONEncoder(encoderCfg), os.Stdout, zap.FatalLevel)
+	return zap.New(core).WithOptions(options...)
+}
+
+func FuzzStep(f *testing.F) {
+	f.Fuzz(func(t *testing.T, data []byte) {
+		initter.Do(initFunc)
+		defer catchPanics()
+		f := fuzz.NewConsumer(data)
```

Comment:
I do not get time to dig into [NewConsumer](github.com/AdaLogics/go-fuzz-headers). Is it similar to the [seed corpus](https://go.dev/security/fuzz/) supported starting from golang 1.18? What's the benefit to use [go-fuzz-headers](github.com/AdaLogics/go-fuzz-headers)? 

Have you considered to use seed corpus directly?  Please see example https://github.com/etcd-io/etcd/blob/2e790d2c4f57eebdb266537ffe9eb7801ebc7cc3/server/etcdserver/api/v3rpc/validationfuzz_test.go#L27-L31

Reply:
`NewConsumer` is not related to the seed corpus feature of golang 1.18 fuzzing. `NewConsumer` is a helper utility to create different types of structured data formats which in this fuzzer is a struct. You can also easily use it to creates slice and/or tar archives.